### PR TITLE
fix(datetime): handle am / pm variants

### PR DESCRIPTION
**Pull request by timreichen**
- allows for am / pm variants as described in http://www.unicode.org/reports/tr35/tr35-dates.html#Date_Field_Symbol_Table

### CODE REVIEW

**Review by iuioiua — COMMENTED** [review]
To my surprise, these tests all pass even without the implementation change. Why? I think we should add more thorough testing for `dayPeriod` before working on this addition. Also, we should update the `format` and `parse` documentation to reflect this change.

**Comment by babiabeo** [comment]
I thunk the problem is in `partsToDate` method:

https://github.com/denoland/std/blob/3af2ed30a98b7fc5076cbb329bcf3e5e415b21ff/datetime/_date_time_formatter.ts#L642-L649

It only checks if `dayPeriod.value` is `”PM”`. Otherwise, it ignores other values. That's why these tests all pass even without the implementation change.

**Comment by iuioiua** [comment]
@babiabeo, would you be interested in creating some tests to compliment this PR?

**Comment by iuioiua** [comment]
@timreichen, reminder that we can't merge this PR until the tests that it touches has been improved. Please see the above comments.

**Comment by timreichen** [comment]
Please hold with this PR until https://github.com/denoland/std/pull/5622.

**Comment by iuioiua** [comment]
@timreichen, have the tests been improved as requested above?

**Comment by timreichen** [comment]
> @timreichen, have the tests been improved as requested above?

Yes, I added tests for all valid cases and a test that throws on invalid data.

File: datetime/_date_time_formatter_test.ts [hunk 1]
```diff
@@ -87,6 +94,143 @@ Deno.test("dateTimeFormatter.partsToDate()", () => {
...
+      { type: "fractionalSecond", value: "000" },
+      { type: "dayPeriod", value: "am." },
+      { type: "timeZoneName", value: "UTC" },
+    ]),
+    +date,
+  );
+});
+Deno.test("dateTimeFormatter.partsToDate() works with dayPeriod", () => {
+  const date = new Date("2020-01-01T00:00:00.000Z");
+  using _time = new FakeTime(date);
```
Comment (iuioiua) [inline datetime/_date_time_formatter_test.ts:142]: Are these lines needed?

File: datetime/_date_time_formatter_test.ts [hunk 2]
```diff
@@ -87,6 +94,143 @@ Deno.test("dateTimeFormatter.partsToDate()", () => {
...
+      { type: "minute", value: "00" },
+      { type: "second", value: "00" },
+      { type: "fractionalSecond", value: "000" },
+      { type: "dayPeriod", value: "am." },
+      { type: "timeZoneName", value: "UTC" },
+    ]),
+    +date,
+  );
+});
+Deno.test("dateTimeFormatter.partsToDate() works with dayPeriod", () => {
```
Comment (iuioiua) [inline datetime/_date_time_formatter_test.ts:140]: Can you please add PM tests?

File: datetime/_date_time_formatter_test.ts [hunk 3]
```diff
@@ -87,6 +94,143 @@ Deno.test("dateTimeFormatter.partsToDate()", () => {
...
+      { type: "hour", value: "00" },
+      { type: "dayPeriod", value: "a.m." },
+      { type: "timeZoneName", value: "UTC" },
+    ]),
+    +date,
+  );
+});
+Deno.test("dateTimeFormatter.partsToDate() throws with invalid dayPeriods", () => {
+  const date = new Date("2020-01-01T00:00:00.000Z");
+  using _time = new FakeTime(date);
```
Comment (iuioiua) [inline datetime/_date_time_formatter_test.ts:196]: Ditto. Are these needed?

File: datetime/_date_time_formatter_test.ts [hunk 4]
```diff
@@ -87,6 +94,143 @@ Deno.test("dateTimeFormatter.partsToDate()", () => {
...
+  const format = "HH a";
+  const formatter = new DateTimeFormatter(format);
+  assertThrows(() =>
```
Comment (iuioiua) [inline datetime/_date_time_formatter_test.ts:199]: Can you please add the remaining arguments to these `assertThrows()` assertions? I.e. error class and message.

**Review by iuioiua — CHANGES_REQUESTED** [review]

File: datetime/_date_time_formatter_test.ts [hunk 5]
```diff
@@ -88,6 +95,169 @@ Deno.test("dateTimeFormatter.partsToDate()", () => {
     +date,
   );
 });
+Deno.test("dateTimeFormatter.partsToDate() works with am dayPeriod", () => {
+  const date = new Date("2020-01-01T00:00:00.000Z");
+  using _time = new FakeTime(date);
+  const format = "HH a";
+  const formatter = new DateTimeFormatter(format);
+  assertEquals(
+    +formatter.partsToDate([
```
Comment (iuioiua) [inline datetime/_date_time_formatter_test.ts:104]: Are these unary pluses needed? If not, please remove.

**Review by iuioiua — CHANGES_REQUESTED** [review]

File: datetime/_date_time_formatter_test.ts (hunk 5 above)
Comment (timreichen) [reply datetime/_date_time_formatter_test.ts]: done

**Review by iuioiua — APPROVED** [review]
Nice! LGTM.

**Pull request merged**